fix(copilot): refreshing copilot tokens

jones-sam opened the pull request:
https://github.com/yetone/avante.nvim/issues/933

When the copilot token expired, the refresh token function would get called but it would happen async so that the completion would return a 401 since it would use still the expired token. The next request would then work because the async call to update it would then be completed. This would happen every time after the token expired, which was quite frequent since it expires every 30 min. 

This PR fixes it so it gets called synchronously, if the token is expired it will update it first before then calling the completion endpoint.

b0o commented:
This has been going back and forth a lot. Running `plenary.curl` synchronously will lock up the UI until the request completes. I think there's a better way to structure the code so we can refresh asynchronously without causing the bug.

See:
- https://github.com/yetone/avante.nvim/issues/809
- https://github.com/yetone/avante.nvim/pull/918
- https://github.com/yetone/avante.nvim/pull/791

jones-sam commented:
Alright, how about this?

- Added ability to do both sync and async requests depending on the context
- Created timer to refresh token every 28 minutes to prevent ever having an outdated token
- Added lockfile so that only one instance of neovim can have a timer to refresh the token so that we don't send unneeded requests
- Added another timer to check if instance of neovim that was refreshing the token was closed, and take it over
- Added synchronous call to request token before calling the completion endpoint to prevent 401 only if the token is expired, however this should rarely happen since it should be upgraded in the background asynchronously before that ever gets triggered. Basically it is just a backup

tomaszferens commented:
Can we merge it? Currently, I have to run the prompts twice because the first one always fails with "token expired" error

aarnphm reviewed (commented):
I can chime in here, I think we should think of a permanent fix other than this

We probably still want to do async refresh without blocking the main UI.

jones-sam commented:
> I can chime in here, I think we should think of a permanent fix other than this
> 
> We probably still want to do async refresh without blocking the main UI.

This should be a permanent fix as long as github keeps their auth process the same. The token gets refreshed async in the background to make sure it never expires, as far as I am aware that doesn't block any UI or cause any other issues. If we do it async right before sending a completion like it is in the main branch now, it is always guaranteed to fail after since the async request is not finished before the completion request, meaning the new token doesn't get used, resulting in a 401. If we do a sync request before, it will always freeze up, waiting for the request to go through. With this implementation, it does it async in the background, so neither of those problems arise.

I've been using my branch of avante since I created this PR and it has been working well, no issues. I think we should merge this soon because I imagine everyone who is using copilot is running into 401s every 30 min. Are there any changes you'd like in this code? I'm not sure what other approach could be made to solve the problem other than this.

tomaszferens commented:
I think what you did @jones-sam is great, much better than status quo we have today.

albankurti commented:
I've been using this PR's implementation for a few days and it works well. It's much better than having to yank and re-paste queries.

aarnphm commented on lua/avante/providers/copilot.lua:
```diff
@@ -81,35 +137,59 @@ end
+    end
+  end
+
+  if async then
```
why can't we just use async here? In what case where we don't use async?

jones-sam replied:
There is a check to see if the token is expired before sending a completion request, and if it is, then it will send a synchronous request to get a new token before sending the completion request. This is because if the token is expired, the completion request will always return a 401. If we send it async in this case, the completion request will also always fail since it won't complete in time for the following request. 

This shouldn't happen often since it gets refreshed in the background to always keep it up-to-date, but sometimes it will be needed if the background request fails due to the internet being down or another error.

aarnphm commented on lua/avante/providers/copilot.lua:
```diff
@@ -162,20 +244,110 @@ M.parse_curl_args = function(provider, code_opts)
+
+-- Register cleanup on Neovim exit
+vim.api.nvim_create_autocmd("VimLeavePre", {
+  callback = function() M.cleanup() end,
```
wait why do we need to register the cleanup here?

afaik once you exited neovim all plugins should be de-registered. Usually cleanup are only meant for during neovim session.

ig we need to remove the lock file.

jones-sam replied:
I added it so the cleanup would run and remove the lockfile when closing neovim to clean things up right away. But you are right, things would still work without this, however the lockfile would remain with the pid of the closed instance. But once another instance checks the lockfile and sees that that pid is no longer running, it will overwrite it and take over. Up to you if you want to keep this or not, I believe it will still work properly with or without this.

aarnphm replied:
Yeah I don't have a big problem with this. (I might be wrong), but from the documentations it seems like it should only be used with shada or sth

```prolog
VimLeavePre	    Before exiting Vim, just before writing the
				.shada file.  This is executed only once,
				if there is a match with the name of what
				happens to be the current buffer when exiting.
				Mostly useful with a "*" pattern. >
				   :autocmd VimLeavePre * call CleanupStuff()
<				Use |v:dying| to detect an abnormal exit.
				Use |v:exiting| to get the exit code.
				Not triggered if |v:dying| is 2 or more.

```

aarnphm reviewed (approved):
Ok I'm good with merging this for now, given that a lot of folks are using copilot.

I will probably look into this a bit more.
